### ISSUE TITLE
reolink B400 Timeout

### ISSUE THREAD

**Issue by sharee993** [thirtythreeforty/neolink#28]
Hello, I have been fighting Reolink b400 for a few days but I didn't manage to do anything:

```
[2020-07-06T14:35:54Z INFO  neolink] Neolink 0.3.0 (unknown commit) release
[2020-07-06T14:35:55Z INFO  neolink] cam: Connecting to camera at 192.168.0.117:9000
[2020-07-06T14:35:55Z INFO  neolink] cam: Connecting to camera at 192.168.0.117:9000
[2020-07-06T14:35:55Z ERROR neolink] Error streaming from camera cam, will retry in 1s: Timeout
[2020-07-06T14:35:55Z INFO  neolink] cam: Connected to camera, starting video stream subStream
[2020-07-06T14:35:56Z INFO  neolink] cam: Connecting to camera at 192.168.0.117:9000
[2020-07-06T14:35:56Z ERROR neolink] Error streaming from camera cam, will retry in 2s: Timeout
```

**Comment by thirtythreeforty** [comment]
Hi - sorry you're having trouble. So the timeout thing looks a lot like #14 - in particular [this comment](https://github.com/thirtythreeforty/neolink/issues/14#issuecomment-648235535) about the system time being set.

Could you try adding the following line to your `cam`'s config section?

```
timeout = { secs = 5, nanos = 0 }
```

This will lengthen the timeout. It looks like it is successfully connecting, so perhaps this will solve it.

**Comment by thirtythreeforty** [comment]
Also note that this is a terrible config file syntax; therefore it is not documented officially and it will likely change in the future (but only after this problem is solved).

**Comment by sharee993** [comment]
I added but unfortunately nothing helped.  camera time is set correctly

**Comment by thirtythreeforty** [comment]
Hmmm. Could you post your config file? (Delete the passwords from it!)

**Comment by hh1209** [comment]
Are you streaming simultaneously from another device (i.e. an NVR, the mobile app, the PC app)? I noticed that once the substreams were added, I was no longer able to stream the HD feed in two places at once.

Sent from Outlook Mobile<https://aka.ms/blhgte>
________________________________

**Comment by sharee993** [comment]
bind = "0.0.0.0"

[[cameras]]
name = "cam1"
username = "admin"
password = ""
address = "192.168.0.117:9000"
format = " H264 "
timeout = { secs = 5, nanos = 0 }

**Comment by thirtythreeforty** [comment]
@kylekartan or @xyphos10: I know you both have B400s. Since support has landed (with `format = "H264"`) have you been able to get your cameras to stream?

**Comment by sharee993** [comment]
I tried without h264 but the effect is the same

**Comment by thirtythreeforty** [comment]
You definitely need the H264 line. Also - I don't think this makes a difference but make sure it's
```
format = "H264"
```
and not
```
format = " H264"
```
(note the leading space).

If that doesn't fix it, could you post a log after setting `RUST_LOG=debug`? On windows:
```
set RUST_LOG=debug
```
On Linux:
```
export RUST_LOG=debug
```

**Comment by kylekartan** [comment]
I may have made a mistake upgrading to the latest available firmware, this is what I get now with v.30

I also double checked the camera is set to the correct time.

Thanks!

Config:
bind = "0.0.0.0"

[[cameras]]
name = "driveway"
username = "admin"
format = "H264"
timeout = { secs = 5, nanos = 0 }
password = ""
address = "192.168.0.47:9000"


[log.txt](https://github.com/thirtythreeforty/neolink/files/4880085/log.txt)

**Comment by thirtythreeforty** [comment]
Ok. So with two cameras experiencing the same issue, I think it's not either of your faults.

What is needed now is a Wireshark packet capture of the official client talking to the camera. Is anyone comfortable and willing to provide one?  Just fire up Wireshark on your computer, hit Record on the interface, and then open the official client and have it start a stream from the camera.  I need to know what it's doing differently than Neolink.

If you send me one, send it privately (thirtythreeforty@gmail.com). I will keep it confidential.

**Comment by kylekartan** [comment]
No problem, I'll do that this afternoon. Thanks!

**Comment by sharee993** [comment]
ok already sent

**Comment by jjtimmins** [comment]
[Log.txt](https://github.com/thirtythreeforty/neolink/files/4880627/Log2.txt)
I get the same timeout with a D400 using Neolink 0.3.0 even with both new lines in my Config.toml file. Sent packet to your email.
format = "H264"
timeout = { secs = 5, nanos = 0 }

**Comment by sharee993** [comment]
Hello, does anyone have any news?

**Comment by Michaelflat1** [comment]
Could the same occur to D800 4K camera (dome one)? all working ok until it just stops working around 6hrs after, before it showed unknown error.. this time it shows authentication error.
I haven't got any idea, but once you connect via reolink client.. works fine. At least for another few hours anyway. Would really rather not use the reolink client as this setup is running off a raspberry pi.. no PC's here.

**Comment by hh1209** [comment]
@Michaelflat1 I have a D800 camera and it's working with no issues. If you get it to initially connect and get a video stream then you shouldn't be having issues. Not sure why you would get authentication issues after initial successful connection. Just to eliminate differences, are you running the latest version of firmware on the D800? 

Also a screenshot or log of the errors would be helpful.

**Comment by Michaelflat1** [comment]
![image](https://user-images.githubusercontent.com/43958840/87030476-90d28080-c1d9-11ea-9f38-33b40eb16e1a.png)

I don't know how to get log output. This is on a raspberry pi. I'm quite new to linux and etc.
I've ran command export RUST_LOG=debug, has this exported a debug file somewhere? I don't know where.

Thanks in advance. I know you can set on the IP camera to reboot it every week, i wonder if i could do this every day. Might stop these issues. Don't think i'll buy another reolink camera in future.. just a normal RTSP camera. 

Also would be nice if you could expose the balanced stream (at the moment there is just clear and fluent... i don't even know how the balanced stream is made, or what it's properties are.. at a guess it's 858x480 and ~20fps. Would be nice to have 720p medium stream, but I think i would need new camera for that.

**Comment by thirtythreeforty** [comment]
I'm still working on D400 logs... I'm in the middle of buying a house (!) so I appreciate patience.

@Michaelflat1 that screenshot is the logs. The interesting thing is the line:
```
thread 'main' panicked at 'assertion failed `(left != right)`
```
That means the main thread crashed. Could you capture a new log with a backtrace (`export RUST_BACKTRACE=1`) and [open a new issue](https://github.com/thirtythreeforty/neolink/issues/new) for this problem?

**Comment by sharee993** [comment]
I found a temporary solution.
I removed timeout = {secs = 5, nanos = 0} from config config enabled neolink. turning off neolink adds timeout = {secs = 5, nanos = 0} to config and turns on neolink in 3-4 time it connects to the camera and works until the next restart.
then again it alone and works proven since yesterday

**Comment by thirtythreeforty** [comment]
@sharee993 I am confused; having trouble understanding what you did. Can you post the config file that works with the D400s?

**Comment by sharee993** [comment]
1. turn on neo-linguists without timeout = {secs = 5, nanos = 0} in config 
2. Turn off neolink and add to config  timeout = {secs = 5, nanos = 0} 
3. Turn on neolink (it doesn't always work the first time usually  3 together)

**Comment by DanieleQ97** [comment]
Hi everyone, I'm having the same issue with a D400.

If it's any help, here is my debug output from neolink:
![Schermata del 2020-07-10 19-49-26](https://user-images.githubusercontent.com/50218939/87184003-7c7ca980-c2e7-11ea-8924-206c8004e920.png)

Adding or not the timeout line in the config changes nothing.

Also, I'm not having any crash from random threads.

Here is my config file:
`bind = "0.0.0.0"

[[cameras]]
name = "TEST"
username = "admin"
#password = "12345678"
address = "192.168.1.123:9000"
#timeout = { secs = 5, nanos = 0 }`

Unfortunately, at the moment, I don't have any Reolink NVR lying around so I don't think I can be of any help with Wireshark.

**Comment by jjtimmins** [comment]
My Blue Iris compatible RLC-410 5MP streams simultaneously on more than one PC with the Reolink Client app (https://reolink.com/ca/software-and-manual/), but the D400s produce the 'timeout' when trying to open them on a second PC (one D400 with the latest firmware and one not). Closing the client app on the 1st PC releases the D400s, then the D400s streams show up on the second Reolink Client app PC. Seems that **D400s could only stream to one ip at a time**. Even with all closed streams, I still get the timeout error with Neolink, though it shows **connected momentarily** (see debug log below). The log show that the timeouts happen after attempting to open the same camera stream twice in a row. **Could this second consecutive connection attempt cause this timeout** in the same way that we get when trying to stream on a second Reolink Client PC?
![image](https://user-images.githubusercontent.com/67901677/87243857-741b9000-c407-11ea-916a-b35a9496d688.png)

**Comment by thirtythreeforty** [comment]
Hm, that's a great point @jjtimmins. Can you add 
```
stream = "mainStream"
```
to that camera's configuration and retry? Currently it attempts to stream both the main and substream (HD/SD basically). If it's true you can only connect one stream at a time, what could be happening is that the subStream connects first, blocking the bigger stream from ever starting.

**Comment by DanieleQ97** [comment]
It works, that was the problem.

Apparently D400 cameras (and probably B400) only allow one stream at time.

**Comment by thirtythreeforty** [comment]
Ha! Great.

Ok. So now we need to teach Neolink about all these nuances. I think that detecting the type of camera will be a prerequisite (#36). Then, clearly announce which streams it's launching (#35).

**Comment by kylekartan** [comment]
Sweet! Adding the _stream = "mainStream"_ worked for me as well. I had some trouble with it erroring so I increased the timeout value to 8 seconds and now it's working so far.

**Comment by jjtimmins** [comment]
Way to go thirtythreeforty, that worked!!  BTW, timeout not needed. Strange that D400's can only deliver one stream because the Reolink Client app shows two streams are available. Not sure how they make the record work with their brand NVRs for continuous recordings on substream and motion detect recording on mainstream. Would be nice to have BlueIris access both streams.
![image](https://user-images.githubusercontent.com/67901677/87254573-83bfc680-c451-11ea-8864-9f649dd5e5c7.png)

**Comment by DanieleQ97** [comment]
@jjtimmins Maybe they just continuosly switch from one stream to the other instead of connecting to both of them at the same time.
Probably something similar that can also be done in Neolink, maybe telling it to switch from one stream to the other with some RTSP request.

Something like:
1. Client send `PLAY rtsp://127.0.0.1:8554/your_camera_name/subStream RTSP/1.0`
2. Neolink connect to _your_camera_name_ substream.
3. Client send `PLAY rtsp://127.0.0.1:8554/your_camera_name/mainStream RTSP/1.0`
4. Neolink close connection to _your_camera_name_ substream and open mainstream.
5. Etc.

That way you can setup a script with something like Motion that by default request Neolink the substream but ask him to switch to the mainstram if some movement is detected.

**Comment by jjtimmins** [comment]
My BlueIris is not connecting the D400 with the settings at the bottom of #11. When I click Okay, the camera initiates the connection and I get this error. Are D400 BlueIris video configs different?
![image](https://user-images.githubusercontent.com/67901677/87327409-07d18700-c502-11ea-8d87-e8fd949a8919.png)

**Comment by hh1209** [comment]
Do you get that error before or after you try to connect in BI?

**Comment by DanieleQ97** [comment]
@jjtimmins If you are talking about this:
> ![BISetup](https://user-images.githubusercontent.com/54970289/86868400-3c74b700-c09a-11ea-99c5-5a51b03ae290.JPG)
You should probably leave empty the subStream path (or set it the same as mainstream) since you have already disabled it in Neolink adding `stream = "mainStream"` in its config file.

That said, I don't think that is requesting the disabled stream that's crashing Neolink.
You should run `set RUST_BACKTRACE=1` as suggested and maybe also `set RUST_LOG=debug` before running Neolink again so that we can see more clearly what's happening.

**Comment by jjtimmins** [comment]
The error only occurs immediately after I click OK after configuring the camera, or if I rightclick the camera window and 'Restart camera', showing the colored bars in the BI screen then a no signal message, but never any stream. I have left out the substream statement as you noted. Rust backtrace produce the attached log. My Gstreamer installation was not pretty, with 'Typical' leaving missing dlls and a reinstallation of the whole 1.16.2 version to resolve.
[Rust_Backtrace.txt](https://github.com/thirtythreeforty/neolink/files/4916470/Rust_Backtrace.txt)

**Comment by jjtimmins** [comment]
Finally resolved by trouble shooting defective GStreamer installation using (1.16.2 runtime installer) at https://gstreamer.freedesktop.org/download/. Uninstalled it, deleted C:\GStreamer directory, rebooted PC, reinstalled full GStreamer 1.16.2, rebooted, ran Neolink in CMD, still several missing DLLs when running Neolink!?? Copied full C:\Gstreamer folder from laptop where it works (not sure how I got GSteamer to install properly there) to this new NAS build, and now it works, all streams for my three D400s!? Anyone else having issues installing Gstreamer on Win 10 pro? (Build 19041)

**Comment by That-Dude** [comment]
I finally got both of my D400 cameras working with BlueIris - thanks so much George.

I'm copying my settings here in case someone finds them useful.

Install G-Streamer using Choco

`choco install gstreamer`

I grabbed the neolink Windows build from [here](https://github.com/thirtythreeforty/neolink/actions/runs/164704687)

Make a folder c:\neolink and copy in the neolink executable

Create "**config.toml**" in the same folder:

```
bind = "0.0.0.0"

[[cameras]]
name = "cam1"
username = "admin"
password = "yourpassword"
address = "192.168.124.230:9000"
stream = "mainStream"
format = "H264"
timeout = { secs = 5, nanos = 0 }

bind = "0.0.0.0"

[[cameras]]
name = "cam2"
username = "admin"
password = "youpassword"
address = "192.168.124.231:9000"
stream = "mainStream"
format = "H264"
timeout = { secs = 5, nanos = 0 }

```

Setup cameras in BlueIris

<img width="855" alt="Screenshot 2020-07-17 at 16 23 43" src="https://user-images.githubusercontent.com/6509533/87804261-85bbc800-c84b-11ea-8854-1923f9c5b1e8.png">

**Comment by thirtythreeforty** [comment]
Hey all - thanks for troubleshooting this. It's good to know it can handle only one stream at a time. When Neolink autodetects camera type, it'll autoconfigure that quirk as well.

Closing this but feel free to ask questions if you have them.

**Comment by QuantumEntangledAndy** [comment]
So I was looking for our protocol and wanted to ask a question about this one stream issue.

Currently we do this:
- Start camera_loop on main
  - connect to camera
  - subscribe to main
  - loop main message and send to rtsp
- Start camera_loop on sub
  - connect to camera
  - subscribe to sub
  - loop sub message and send to rtsp

Could we instead do
- Start camera_loop
  - connect to camera
  - subscribe to main
  - subscribe to sub
  - loop main and sub message and send to rtsp

**Comment by QuantumEntangledAndy** [comment]
I also notice that on the official client when accessing the substream the handle is 256 but we always use 0. Perhaps we should adjust this and see if it helps with multi access
